src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIListOfTConverter.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Converter for <cref>System.Collections.Generic.IList{TElement}</cref>.
+    /// </summary>
+    internal sealed class JsonIListOfTConverter<TCollection, TElement> : JsonIEnumerableDefaultConverter<TCollection, TElement>
+        where TCollection : IList<TElement>
+    {
+        protected override void Add(TElement value, ref ReadStack state)
+        {
+            ((TCollection)state.Current.ReturnValue!).Add(value);
```

Comment:
When `TCollection` is a value type (that implements `IList<T>`), this unboxing operation creates a copy of the `TCollection` instance and calls the `Add` method in the copy, which prevents that the `value` gets added to the actual boxed `TCollection` in `state.Current.ReturnValue`.
See scenario on https://github.com/dotnet/runtime/issues/1998.

A possible fix could be to use `as` to prevent unboxing (I think) and that way we add to the right instance. However you need to cast to `IList<TElement>` which is kinda redundant to the generic constraint.
```suggestion
            (state.Current.ReturnValue as IList<TElement>)!.Add(value);
```

Reply:
Since we have a tracking issue but no test yet, I think it is best to defer the fix for that issue instead of adding one more area to review here since this PR is too large already.